docs/howtos/when-to-use-what-in-the-ffi.md

```diff
@@ -0,0 +1,133 @@
+
+# When to use what method of passing data between Rust and Java/Swift
+
+There are a bunch of options here. For the purposes of our discussion,
+there are two kinds of values you may want to pass over the FFI.
+
+1. Types with identity
+2. Plain ol' data.
+
+## Types with identity
+
+Examples of this are things like database connections, the FirefoxAccounts
+struct, etc. These types are complex, implemented in rust, and it's not
+unreasonable for them to come to Java/Kotlin as a type representing a
+resource (e.g. implementing `Closable`/`AutoClosable`).
+
+You have two choices here:
+
+1. Use a `ConcurrentHandleMap` to store all instances of your object, and
+   pass the handle back and forth as a u64 from Rust / Long from Kotlin.
+
+   This is recommended for most cases, as it's the hardest to mess up.
+   Additionally, for types T such that `&T: Sync + Send`, or that you
+   need to call `&mut self` method, this is the safest choice.
+
+   Additionally, this will ensure panic-safety, as you'll poison your Mutex.
+
+   The [`ffi_support::handle_map` docs](https://docs.rs/ffi-support/*/ffi_support/handle_map/index.html)
```

Comment:
In markdown, you can define a reference-style link to avoid having to put in the full link each time at the bottom of the file, eg, https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#links